transport-native-epoll/src/main/c/netty_epoll_linuxsocket.c

```diff
@@ -262,6 +267,134 @@ static jobject netty_epoll_linuxsocket_getPeerCredentials(JNIEnv *env, jclass cl
+    struct ifreq ifr;
+    const char *f_ifName = (*env)->GetStringUTFChars(env, ifName, NULL);
+
+    // Initialize the IFREQ structure with the specified flags and interface name.
+    memset(&ifr, 0, sizeof(ifr));
+    ifr.ifr_flags = ((isTapDev) ? IFF_TAP : IFF_TUN);
+    strncpy(ifr.ifr_name, f_ifName, IFNAMSIZ);
+
+    // Open/allocate the TUN/TAP device.  Raise an IOException if an error occurs.
+    res = ioctl(cloneDevFD, TUNSETIFF, (void *)&ifr);
+    if (res < 0) {
+        netty_unix_errors_throwChannelExceptionErrorNo(env, "allocTunTapInterface() failed: ", errno);
+    }
+
+    (*env)->ReleaseStringUTFChars(env, ifName, f_ifName);
+
+    return res;
+}
+
+static jlong netty_epoll_linuxsocket_readTunTapPacketAddress(JNIEnv *env, jclass cls, jint fd, jlong addr, jint pos, jint limit) {
+    jlong res = -1;
+    struct iovec iov[2];
+    struct tun_pi packetInfo;
+
+    iov[0].iov_base = (void *) &packetInfo;
+    iov[0].iov_len = sizeof(packetInfo);
+    iov[1].iov_base = ((char *) addr) + pos;
+    iov[1].iov_len = limit;
+
+    // Attempt to read a packet. Loop if the read is interrupted by a signal.
+    do {
+        res = (jlong) readv((int) fd, iov, 2);
+    } while (res < 0 && errno == EINTR);
+
+    // If a packet was received...
+    if (res >= (jlong) sizeof(packetInfo)) {
+
+        // If the packet was truncated because the receive buffer was too small,
+        // raise an IOException containing the error ENOMEM.
+        if ((packetInfo.flags & TUN_PKT_STRIP) != 0) {
+            netty_unix_errors_throwIOExceptionErrorNo(env, "readTunTapPacketAddress() failed: ", ENOMEM);
+            res = -1;
+        }
+
+        // Otherwise return the protocol type in the upper 16 bits and the length
+        // of the packet (not including the packet info header) in the lower 32 bits.
+        else {
+            jlong proto = htons(packetInfo.proto);
+            jlong packetLen = res - sizeof(packetInfo);
+            res = (proto << 32) | packetLen;
+        }
+    }
+
+    // If an error occurred, return to the caller empty-handed if the kernel signaled
+    // that no packet was available.  Otherwise, raise an IOException containing the error code.
+    else if (res < 0) {
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+            res = 0;
+        }
+        else {
+            netty_unix_errors_throwIOExceptionErrorNo(env, "readTunTapPacketAddress() failed: ", errno);
+        }
+    }
+
+    // If for some reason the kernel returned less bytes than are needed to
+    // convey the packet info header, treat it as if no packet had been received.
+    // (Note that it's not clear the kernel will ever do this in practice,
+    // therefore this code is purely defensive).
+    else {
+        res = 0;
+    }
+
+    return res;
+}
+
+static jlong netty_epoll_linuxsocket_readTunTapPacket(JNIEnv *env, jclass cls, jint fd, jobject jbuf, jint pos, jint limit) {
+    void *buf = (*env)->GetDirectBufferAddress(env, jbuf);
```

Comment:
It looks like this may be called when the buffer doesn't have an address ... I made a suggestion/comment in he java code on this topic about how we typically handle this case.